### DISCUSSION TITLE
G96 ( Constant Surface Speed ) on Lathe mode is idle run

### DISCUSSION THREAD

**Discussion by Reprosense** [grblHAL/core#403] [Q&A]
https://github.com/grblHAL/core/assets/135573322/f295cf1b-ca45-48b8-8136-b1929fdcb9ac

hi guys , i just test a G96 on my lathe machine , but it seems that the Machine status is in idle run conditions when its actually run the gcode , can i have a clue what is happening , is it because it should be count the spindle rpm before it run again ? i try to figure it out but i have no clue , here is the video and the G96 ( CSS) code

    if (command_words.G14) {
        if(gc_block.modal.spindle.rpm_mode == SpindleSpeedMode_CSS) {
            if(!gc_state.spindle.hal->cap.variable)
                FAIL(Status_GcodeUnsupportedCommand);
            if (!gc_block.words.s) // TODO: add check for S0?
                FAIL(Status_GcodeValueWordMissing);
    // see below!! gc_block.values.s *= (gc_block.modal.units_imperial ? MM_PER_INCH * 12.0f : 1000.0f); // convert surface speed to mm/min
            if (gc_block.words.d) {
                gc_state.spindle.hal->param->css.max_rpm = min(gc_block.values.d, gc_state.spindle.hal->rpm_max);
                gc_block.words.d = Off;
            } else
                gc_state.spindle.hal->param->css.max_rpm = gc_state.spindle.hal->rpm_max;
        } else if(gc_state.modal.spindle.rpm_mode == SpindleSpeedMode_CSS) {
            if(gc_state.spindle.css) {
                gc_state.spindle.css = NULL;
                protocol_buffer_synchronize(); // Empty planner buffer to ensure we get RPM at end of last CSS motion
            }
            gc_state.spindle.rpm = gc_state.spindle.hal->param->rpm; // Is it correct to restore latest spindle RPM here?
        }
        gc_state.modal.spindle.rpm_mode = gc_block.modal.spindle.rpm_mode;
    }

**Comment by terjeio** [comment]
Post the gcode you are running. And also the $I output from the controller.

Here is a test file I am using - and it seems to work as it should:
[g96.txt](https://github.com/grblHAL/core/files/13643373/g96.txt)

**Comment by Reprosense** [comment]
[VER:1.1f.20231005:]
[OPT:VNZHSL,35,1024,3,0]
[AXS:3:XYZ]
[NEWOPT:ENUMS,RT+,HOME,NOPROBE,ES,LATHE,SS,SED,PID]
[FIRMWARE:grblHAL]
[NVS STORAGE:*FLASH]
[FREE MEMORY:32K]
[DRIVER:STM32F446]
[DRIVER VERSION:230922]
[BOARD:Protoneer v3]
[AUX IO:1,2,0,0]
ok

here , i give u the gcode ([facing.txt](https://github.com/grblHAL/core/files/13644505/facing.txt))

**Comment by Reprosense** [comment]
// Empty planner buffer to ensure spindle is set when programmed.
        if((ok = protocol_buffer_synchronize()) && set_state(spindle, state, rpm) && !at_speed) {
            float on_delay = 0.0f;
            while(!(at_speed = spindle->get_state().at_speed)) {
              //  delay_sec(0.2f, DelayMode_Dwell);
               // on_delay += 0.2f;
                if(ABORTED)
                    break;
                if(on_delay >= settings.safety_door.spindle_on_delay) {
                    gc_spindle_off();
                    system_raise_alarm(Alarm_Spindle);
                    break;
                }
            }
        }

        ok &= at_speed;
    }

    return ok;
}

its in spindle control.c , im commenting the delay_sec & on_delay and now its work perfectly fine without idle run , but did the delay really needed ?

**Reply by terjeio** [reply]
Some delay is needed to avoid spamming/choking VFD spindles with Modbus messages, this since status requests are sent [without waiting for a reply](https://github.com/grblHAL/Plugins_spindle/blob/2eebb9e234fe3db26fa20ff43445a10360a7befa/vfd/gs20.c#L139-L162). The VFD driver code _could_ be changed to simply drop requests if one is pending, perhaps I'll try that later. It is on my todo list.
